### PR TITLE
Add variant struct code-gen

### PR DESCRIPTION
**Pull request by lovio**
This PR will generate a `std::variant` using alias and corresponding `<<` operator: 
``` cpp
template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };

using V = std::variant<std::string, int32_t, double, ::variants_ptr::Simple, ::variants_ptr::Simple*>;

std::ostream& operator<<(std::ostream& stream, const V& value)
{
    std::visit(
        overloaded
        {
            [&stream](const std::string& v) { stream << v; }
            , [&stream](int32_t v) { stream << v; }
            , [&stream](double v) { stream << v; }
            , [&stream](const ::variants_ptr::Simple& v) { stream << v; }
            , [&stream](::variants_ptr::Simple* v) { stream << *v; }
        },
        value);
    return stream;
}
```

Then use `V` in struct

### CODE REVIEW

File: source/generator_cpp.cpp [hunk 1]
```diff
@@ -377,6 +378,19 @@ auto assign_member([[maybe_unused]] Alloc alloc) -> T {
     Write(code);
 }
 
+void GeneratorCpp::GenerateVariantVisitHelper_Header()
+{
+    std::string code = R"CODE(
+template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
```
[the remark this answers is not in the document]
Comment (lovio) [reply source/generator_cpp.cpp:384]: 只用了第一个，我改一下

[the remark this answers is not in the document]
Comment (hurricane1026) [reply source/generator_cpp.cpp:384]: 这是类型递归么？

[the remark this answers is not in the document]
Comment (lovio) [reply source/generator_cpp.cpp:384]: 也不算是递归吧，就是一个参数pack and fold的trick

[the remark this answers is not in the document]
Comment (lovio) [reply source/generator_cpp.cpp:384]: https://www.modernescpp.com/index.php/smart-tricks-with-fold-expressions

[the remark this answers is not in the document]
Comment (lovio) [reply source/generator_cpp.cpp:384]: done

**Review by hurricane1026 — APPROVED** [review]

**Pull request merged**